### PR TITLE
feat: [IDP-532]: Support for OAuth based plugins

### PR DESCRIPTION
**Pull request by vigneswara-propelo**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48390)

### CODE REVIEW

**Comment by vigneswara-propelo** [comment]
trigger uts

**Comment by vigneswara-propelo** [comment]
trigger uts

**Review by mankrit-singh — DISMISSED** [review]

File: idp-service/src/main/resources/configs/auth/json-schemas/github-auth-schema.json [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  "type": "object",
+  "additionalProperties": false,
+  "required": [
+    "auth"
+  ],
+  "properties": {
+    "auth": {
+      "$id": "#root/auth",
+      "title": "Auth",
+      "type": "object",
```
Comment (devesh-harness) [inline idp-service/src/main/resources/configs/auth/json-schemas/github-auth-schema.json:15]: "additionalProperties": false
add this property in every object

File: idp-service/src/main/resources/configs/auth/json-schemas/github-auth-schema.json [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+          "$id": "#root/auth/environment",
+          "title": "Environment",
+          "type": "string",
+          "default": "",
+          "pattern": "^.*$"
+        },
+        "providers": {
+          "$id": "#root/auth/providers",
+          "title": "Providers",
+          "type": "object",
```
Comment (devesh-harness) [inline idp-service/src/main/resources/configs/auth/json-schemas/github-auth-schema.json:31]: "additionalProperties": false
add this property in every object

File: idp-service/src/main/resources/configs/auth/json-schemas/github-auth-schema.json [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+          "required": [
+            "github"
+          ],
+          "properties": {
+            "github": {
+              "$id": "#root/auth/providers/github",
+              "title": "Github",
+              "type": "object",
+              "required": [
```
Comment (devesh-harness) [inline idp-service/src/main/resources/configs/auth/json-schemas/github-auth-schema.json:40]: "additionalProperties": false
add this property in every object

File: idp-service/src/main/resources/configs/auth/json-schemas/github-auth-schema.json [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                "development"
+              ],
+              "properties": {
+                "development": {
+                  "$id": "#root/auth/providers/github/development",
+                  "title": "Development",
+                  "type": "object",
```
Comment (devesh-harness) [inline idp-service/src/main/resources/configs/auth/json-schemas/github-auth-schema.json:47]: "additionalProperties": false
add this property in every object

File: idp-service/src/main/resources/configs/auth/json-schemas/google-auth-schema.json [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  "type": "object",
+  "additionalProperties": false,
+  "required": [
+    "auth"
+  ],
+  "properties": {
+    "auth": {
+      "$id": "#root/auth",
+      "title": "Auth",
+      "type": "object",
```
Comment (devesh-harness) [inline idp-service/src/main/resources/configs/auth/json-schemas/google-auth-schema.json:15]: "additionalProperties": false
add this property in every object

File: idp-service/src/main/resources/configs/auth/json-schemas/google-auth-schema.json [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+          "$id": "#root/auth/environment",
+          "title": "Environment",
+          "type": "string",
+          "default": "",
+          "pattern": "^.*$"
+        },
+        "providers": {
+          "$id": "#root/auth/providers",
+          "title": "Providers",
+          "type": "object",
```
Comment (devesh-harness) [inline idp-service/src/main/resources/configs/auth/json-schemas/google-auth-schema.json:31]: "additionalProperties": false
add this property in every object

File: idp-service/src/main/resources/configs/auth/json-schemas/google-auth-schema.json [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+          "required": [
+            "google"
+          ],
+          "properties": {
+            "google": {
+              "$id": "#root/auth/providers/google",
+              "title": "Google",
+              "type": "object",
+              "required": [
```
Comment (devesh-harness) [inline idp-service/src/main/resources/configs/auth/json-schemas/google-auth-schema.json:40]: "additionalProperties": false
add this property in every object

File: idp-service/src/main/resources/configs/auth/json-schemas/google-auth-schema.json [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                "development"
+              ],
+              "properties": {
+                "development": {
+                  "$id": "#root/auth/providers/google/development",
+                  "title": "Development",
+                  "type": "object",
```
Comment (devesh-harness) [inline idp-service/src/main/resources/configs/auth/json-schemas/google-auth-schema.json:47]: "additionalProperties": false
add this property in every object

**Review by devesh-harness — CHANGES_REQUESTED** [review]

File: idp-service/src/main/java/io/harness/idp/configmanager/ConfigType.java [hunk 9]
```diff
@@ -13,5 +13,6 @@
 @OwnedBy(HarnessTeam.IDP)
 public enum ConfigType {
   PLUGIN,
-  INTEGRATION;
+  INTEGRATION,
+  AUTH;
```
Comment (devesh-harness) [inline idp-service/src/main/java/io/harness/idp/configmanager/ConfigType.java:17]: Here if we are adding new type then on configuration screen where we show the combined configurations for enabled plugins then in that we are not planning to show auth configurations there?

Comment (vigneswara-propelo) [reply idp-service/src/main/java/io/harness/idp/configmanager/ConfigType.java:17]: yes, we are not planning to show auth configurations there. We will have a separate screen to show the clientId and clientSecret for users to configure.

Comment (devesh-harness) [reply idp-service/src/main/java/io/harness/idp/configmanager/ConfigType.java:17]: ok

**Review by devesh-harness — APPROVED** [review]

**Comment by vigneswara-propelo** [comment]
trigger uts

**Comment by vigneswara-propelo** [comment]
trigger uts

**Comment by vigneswara-propelo** [comment]
trigger copyrightcheck

**Comment by vigneswara-propelo** [comment]
trigger ut2

**Review by sandeepa-kv-jois — CHANGES_REQUESTED** [review]
Few things we need to consider:
1) Provide relevant IDP URL values that customers need to configure like Authorized JavaScript origins and  Authorized Redirect URIs in API response
2) Provide relevant IDP URL values that customers need to configure for callbackUrl in case of Github OAuth 
3) enterpriseInstanceUrl if needed or not check based on the connector gitHub Type(GitHub Vs GHE)

**Comment by vigneswara-propelo** [comment]
> 

@sandeepa-kv-jois - Connected with UI team and we decided Authorized JavaScript origins and Redirect URIs will be constructed from UI itself. We just need to give namespace info in the API response.

**Comment by vigneswara-propelo** [comment]
trigger uts

**Review by sandeepa-kv-jois — APPROVED** [review]

**Comment by vigneswara-propelo** [comment]
trigger uts

File: idp-service/contracts/openapi/v1/openapi.yaml [hunk 10]
```diff
@@ -1663,6 +1663,30 @@ components:
...
+      title: AuthInfo
+      x-stoplight:
+        id: v6t3w7hydbjp3
+      type: object
+      properties:
+        namespace:
+          x-stoplight:
+            id: fp9ttpkqvfppc
+          type: string
+        envVariables:
```
Comment (mankrit-singh) [inline idp-service/contracts/openapi/v1/openapi.yaml:1684]: env_variables

Comment (vigneswara-propelo) [reply idp-service/contracts/openapi/v1/openapi.yaml]: done

**Review by mankrit-singh — APPROVED** [review]

**Review by sandeepa-kv-jois — APPROVED** [review]

**Comment by vigneswara-propelo** [comment]
trigger review

**Pull request merged**